File: app_codebase/dart_bkend_base_02/bin/core/ws/ws_app.dart
/// Mount WebSocket tier endpoints and combine with HTTP handler.
library;

import 'package:shelf/shelf.dart';
import 'package:shelf_web_socket/shelf_web_socket.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

import '../auth/auth_config.dart';
import '../errors/module_error_registry.dart';
import '../http/middleware/error_middleware.dart';
import '../http/service/routes.dart';
import '../../modules/module_registry.dart';
import 'service/channel_registry.dart';
import 'ws_dispatcher.dart';

Handler createCombinedHandler() {
  requireSecretsForProduction();
  resetModuleErrorRegistry();
  registerApplicationErrors();
  resetRouteRegistry();
  registerApplicationRoutes();
  registerApplicationState();
  resetChannelRegistry();
  registerApplicationWsChannels();

  final httpHandler = Pipeline()
      .addMiddleware(errorMiddleware())
      .addHandler(buildApplicationHandler());
  final wsHandler = _buildWsHandler();

  return (Request request) async {
    final path = request.requestedUri.path;
    if (path.startsWith('/ws/')) {
      return wsHandler(request);
    }
    return httpHandler(request);
  };
}

Handler _buildWsHandler() {
  Handler mount(String tier) {
    return webSocketHandler((WebSocketChannel channel, _) {
      runWsConnection(channel, tier: tier);
    });
  }

  final public = mount('public');
  final authuser = mount('authuser');
  final service = mount('service');

  return (Request request) async {
    final path = request.requestedUri.path;
    if (path == '/ws/public') return public(request);
    if (path == '/ws/authuser') return authuser(request);
    if (path == '/ws/service') return service(request);
    return Response.notFound('WebSocket endpoint not found');
  };
}

File: app_codebase/python_base_05/bin/core/ws/ws_app.py
"""Attach WebSocket routes to the FastAPI app."""

from __future__ import annotations

from fastapi import FastAPI, WebSocket

from core.errors.module_error_registry import reset_module_error_registry
from core.ws.service.channel_registry import reset_channel_registry
from core.ws.ws_dispatcher import run_ws_connection
from modules.module_registry import register_application_channels, register_application_errors


def configure_websockets(app: FastAPI) -> None:
    """Register WS channel modules and mount /ws/* routes on ``app``."""
    reset_module_error_registry()
    register_application_errors()
    reset_channel_registry()
    register_application_channels()

    @app.websocket("/ws/public")
    async def ws_public(websocket: WebSocket) -> None:
        await run_ws_connection(websocket, tier="public")

    @app.websocket("/ws/authuser")
    async def ws_authuser(websocket: WebSocket) -> None:
        await run_ws_connection(websocket, tier="authuser")

    @app.websocket("/ws/service")
    async def ws_service(websocket: WebSocket) -> None:
        await run_ws_connection(websocket, tier="service")
